yew/src/virtual_dom/vlist.rs

```diff
@@ -78,137 +173,320 @@ impl VDiff for VList {
+                    if !hash_set.insert(k) {
+                        log::error!("Key '{}' is not unique in list but must be.", &k);
                     }
                 }
             }
+            hash_set.len()
+        };
+        let some_keyed = n_keyed_lefts > 0;
+        let all_keyed = self.children.len() == n_keyed_lefts;
+        if some_keyed && !all_keyed {
+            log::error!(
+                "Not all elements have keys in VList ({} keyed out of {}), this is currently not \
+                supported. Ignoring keys.",
+                n_keyed_lefts,
+                self.children.len(),
+            );
         }
 
-        // Process children
-        let lefts = self.children.iter_mut();
-        let key_count = rights.iter().filter(|r| r.key().is_some()).count();
-        let mut rights_nokeys = Vec::with_capacity(rights.len() - key_count);
-        let mut rights_lookup = HashMap::with_capacity(key_count);
-        for mut r in rights.into_iter() {
-            if let Some(key) = r.key() {
-                if rights_lookup.contains_key(key) {
-                    log::error!("Duplicate key of {}", &key);
-                    r.detach(parent);
+        // Take rights from ancestor.
+        let mut rights = match ancestor {
+            // If the ancestor is also a VList, then the right list is the
+            // previously rendered items.
+            Some(VNode::VList(vlist)) => vlist.children,
+            // Otherwise, there was a node before but it wasn't a list. Then,
+            // use the current node as a single fragment list and let the
+            // `apply` of `VNode` handle it.
+            Some(vnode) if !some_keyed => vec![vnode],
+            // Otherwise, we don't reuse it, as the chance that the element
+            // is keyed and present in left is almost null.
+            Some(mut vnode) => {
+                vnode.detach(parent);
+                vec![]
+            }
+            None => vec![],
+        };
+
+        // Collect the right/ancestor keyed elements in a hashmap.
+        let mut rights_lookup = rights
+            .iter()
+            .enumerate()
+            .filter_map(|(idx, vnode)| vnode.key().as_ref().map(|key| (key.clone(), idx)))
+            .collect::<HashMap<_, _>>();
+
+        // Determine which algorithm we use. If there are some keys, but not all
+        // the elements are keyed, we consider it a degenerated case and use the
+        // non-keyed algorithm. Importantly, don't use the keyed algorithm if
+        // rights is empty (no need).
+        let all_rights_have_keys = rights_lookup.len() == rights.len();
+        let use_keyed_algorithm =
+            some_keyed && all_keyed && all_rights_have_keys && !rights.is_empty();
+
+        // The algorithms are different when there are keys, because it is more
+        // expensive and less frequent.
+        if !use_keyed_algorithm {
+            let mut rights = rights.into_iter();
```

Comment:
I think we need to be careful not to reuse keyed `rights` in the non keyed algorithm

Reply:
You're right, I forgot about that. Then I propose to just change the definition of `use_keyed_algorithm` to `some_keyed && all_keyed && all_rights_have_keys` (IMO if there are keys, all vnodes should have keys, so it's ok to break that contract if this is not the case. Otherwise the diffing algorithm is hell to implement and cover all cases. Or I just need some help :slightly_smiling_face: ).

Also, I have difficulties with writing tests that use `VComp`s. I failed to reproduce an error that I had in the example using tests. So I would appreciate if you could propose one/some test(s) showcasing this :slightly_smiling_face:

Reply:
> Then I propose to just change the definition of use_keyed_algorithm to some_keyed && all_keyed && all_rights_have_keys

I think I know what you're proposing, did you mean `|| all_rights_have_keys`?

> Also, I have difficulties with writing tests that use VComps. I failed to reproduce an error that I had in the example using tests. So I would appreciate if you could propose one/some test(s) showcasing this 🙂

Sure! Happy to help

Reply:
This should now be fixed. I didn't write tests though.